estimatey-api/Estimatey.Infrastructure/DevOps/LinksSynchronizationService.cs

```diff
@@ -26,17 +29,28 @@ internal class LinksSynchronizationService
 
     public async Task SyncProject(ProjectEntity project)
     {
-        _logger.LogInformation("Syncing links for project {projectName}.", project.DevOpsProjectName);
+        var semaphore = _semaphores.GetOrAdd(project.Id, _ => new SemaphoreSlim(1, 1));
+
+        await semaphore.WaitAsync();
+
+        try
+        {
+            _logger.LogInformation("Syncing links for project {projectName}.", project.DevOpsProjectName);
 
-        var workItemLinksBatch = await _devOpsClient.GetWorkItemLinksBatch(project.DevOpsProjectName, project.LinksContinuationToken)
-;
-        await ProcessWorkItemLinks(project, workItemLinksBatch);
+            var workItemLinksBatch = await _devOpsClient.GetWorkItemLinksBatch(project.DevOpsProjectName, project.LinksContinuationToken)
+    ;
+            await ProcessWorkItemLinks(project, workItemLinksBatch);
 
-        _logger.LogInformation("Successfully synced links for project {projectName}.", project.DevOpsProjectName);
+            _logger.LogInformation("Successfully synced links for project {projectName}.", project.DevOpsProjectName);
 
-        if (!workItemLinksBatch.IsLastBatch)
+            if (!workItemLinksBatch.IsLastBatch)
+            {
+                await SyncProject(project);
+            }
+        }
+        finally
         {
-            await SyncProject(project);
+            semaphore.Release();
```

Comment:
I believe you'll want to also remove the semaphore from the `ConcurrentDictionary` e.g. ` _semaphoreSlims.TryRemove(project.Id, out _);`

Reply:
Why, can't they be re-used?

Reply:
Yeah, but it's just so the unused semaphore isn't sitting around in memory.

Reply:
Not sure that is necessary, not much memory.  I'm just gonna leave it.